rust/cjdns_sys/src/interface/wire/message.rs

```diff
@@ -0,0 +1,145 @@
+//! Message type.
+
+use thiserror::Error;
+
+/// A network message.
+///
+/// Can be converted to/from to C `struct Message` from `wire/Message.h`.
+#[derive(Clone)]
+pub struct Message {
+    /// Actual data starts at offset `padding`.
+    data: Box<[u8]>,
+
+    /// The number of bytes of padding BEFORE the actual message data.
+    /// Useful to prepend additional data at no cost.
+    padding: usize,
+}
+
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+pub enum MessageError {
+    #[error("Can't prepend {0} bytes, only {1} bytes capacity available")]
+    InsufficientPadding(usize, usize),
+
+    #[error("Can't take {0} bytes from the message, only {1} bytes available")]
+    InsufficientLength(usize, usize),
+}
+
+pub type Result<T> = std::result::Result<T, MessageError>;
+
+impl Message {
+    /// Create new message with zero-initialized content of a specified size.
+    /// The `prepend_capacity` bytes is added to the message capacity *before* the actual data,
+    /// so that the message can be easily prepended with more data.
+    #[inline]
+    pub fn new(size: usize, prepend_capacity: usize) -> Self {
+        Message {
+            data: vec![0_u8; size + prepend_capacity].into_boxed_slice(),
+            padding: prepend_capacity,
+        }
+    }
+
+    /// Get the message length.
+    #[inline]
+    pub fn len(&self) -> usize {
+        debug_assert!(self.data.len() >= self.padding);
+        self.data.len() - self.padding
+    }
+
+    /// Get the available padding size.
+    #[inline]
+    pub fn pad(&self) -> usize {
+        self.padding
+    }
+
+    /// Get the read-only view into the message as byte slice.
+    #[inline]
+    pub fn bytes(&self) -> &[u8] {
+        debug_assert!(self.padding <= self.data.len());
+        &self.data[self.padding..]
+    }
+
+    /// Get the mutable view into the message as byte slice.
+    #[inline]
+    pub fn bytes_mut(&mut self) -> &mut [u8] {
+        debug_assert!(self.padding <= self.data.len());
+        &mut self.data[self.padding..]
+    }
+
+    /// Push additional data `bytes` *before* the message's existing data.
+    /// The available padding must be enough to accommodate additional data,
+    /// otherwise error is returned.
+    pub fn push(&mut self, bytes: &[u8]) -> Result<()> {
+        if bytes.len() > self.padding {
+            return Err(MessageError::InsufficientPadding(bytes.len(), self.padding));
+        }
+        self.padding -= bytes.len();
+        debug_assert!(self.padding <= self.data.len());
+        let dest = &mut self.data[self.padding..self.padding + bytes.len()];
+        dest.copy_from_slice(bytes);
+        Ok(())
+    }
+
+    /// Pop specified number of bytes from the beginning of the message.
+    /// The message must be big enough, otherwise error is returned.
+    pub fn pop(&mut self, count: usize) -> Result<Vec<u8>> {
```

Comment:
I'd rather have:
```rust
pub fn pop<T>(&mut self) -> Result<T>;
pub fn push<T>(&mut self, t: &T) -> Result<()>;

for X in u16, u32, u64
  pub fn popX_le(&mut self) -> Result<uX>;
  pub fn pushX_le(&mut self, t: uX) -> Result<()>;
  pub fn popX_be(&mut self) -> Result<uX>;
  pub fn pushX_be(&mut self, t: uX) -> Result<()>;
```

These match the functions in Message.h and will be easier to work with

Reply:
That makes sense. See the update.